v3/src/components/graph/adornments/adornments.tsx

```diff
@@ -12,16 +13,37 @@ import { useGraphDataConfigurationContext } from "../hooks/use-graph-data-config
 import { useGraphLayoutContext } from "../hooks/use-graph-layout-context"
 import { getAdornmentComponentInfo } from "./adornment-component-info"
 import { updateCellKey } from "./adornment-utils"
+import { kGraphAdornmentsBannerHeight } from "./adornment-types"
 
 import "./adornments.scss"
 
 export const Adornments = observer(function Adornments() {
-  const graphModel = useGraphContentModelContext(),
-    dataConfig = useGraphDataConfigurationContext(),
-    instanceId = useInstanceIdContext(),
-    layout = useGraphLayoutContext(),
-    { isTileSelected } = useTileModelContext(),
-    adornments = graphModel.adornmentsStore.adornments
+  const graphModel = useGraphContentModelContext()
+  const dataConfig = useGraphDataConfigurationContext()
+  const instanceId = useInstanceIdContext()
+  const layout = useGraphLayoutContext()
+  const { isTileSelected } = useTileModelContext()
+  const adornments = graphModel.adornmentsStore.adornments
+  const bannerCount = graphModel.adornmentsStore.activeBannerCount
+  const { left, top, width, height } = layout.computedBounds.plot
+  const defaultTop = useRef<number>(Math.max(top - bannerCount * kGraphAdornmentsBannerHeight, 0))
+  
+  useEffect(function handleAdornmentBannerCountChange() {
+    const updateLayout = () => {
+      const bannersHeight = bannerCount * kGraphAdornmentsBannerHeight
+      const topWithBanners = defaultTop.current + bannersHeight
+      layout.setDesiredExtent("top", topWithBanners)
+    }
+  
+    updateLayout()
+  
+    return () => {
+      mstAutorun(
+        updateLayout,
+        { name: "Graph.handleAdornmentBannerCountChange" }, graphModel
+      )
+    }
+  }, [layout, graphModel, bannerCount])
```

Comment:
I think what you want here is this simpler version:
```suggestion
  const { left, top, width, height } = layout.computedBounds.plot

  useEffect(function handleAdornmentBannerCountChange() {
    return mstAutorun(() => {
      const bannerCount = graphModel.adornmentsStore.activeBannerCount
      const bannersHeight = bannerCount * kGraphAdornmentsBannerHeight
      layout.setDesiredExtent("top", bannersHeight)
      }, { name: "Graph.handleAdornmentBannerCountChange" }, graphModel
    )
  }, [graphModel, layout])
```
There are several issues with your version:
1. the `defaultTop` ref will be computed on every render, but only the the first value will be retained by the ref.
2. the purpose of the `mstAutorun` is apparently to recompute the banner height when the number of banners changes, but the `bannerCount` is retrieved outside the `mstAutorun`, so changes won't trigger it.
3. the return value of the effect is a function which calls `mstAutorun`, which means that when the effect's cleanup function is run the autorun will be installed but will never be cleaned up.
4. when the `bannerCount` changes, the effect will run, which will a) execute the cleanup function, installing an autorun that will execute `updateLayout`, b) call `updateLayout` again as part of the effect, and c) return a new cleanup function.

With my version, the effect is only re-run when the `layout` or the `graphModel` change, which is effectively never. The `mstAutorun` accesses all of the observables it cares about, primarily the `bannerCount`, so it will re-run on change. And the return value of the effect is a cleanup function which disposes of the autorun, because that's the return value of `mstAutorun`. Finally, I don't think it's necessary to subtract the banner height from the top and then add it back later -- I'm pretty sure the top value we want is just the total banner height. Check me to make sure that this all still works.

Reply:
@kswenson Thank 👍 As discussed elsewhere, I made the change you suggested and created [a new story about addressing the top split issue](https://www.pivotaltracker.com/story/show/186864009).